### ISSUE TITLE
Module 3: Training Loop - AMP & Checkpointing (2 Pomodoros)

### ISSUE THREAD

**Issue by gregory1506** [gregory1506/arc-pytorch-prep#3]
## Module 3: Training Loop - AMP & Checkpointing
**Duration**: 2 Pomodoros (50 minutes)  
**Status**: 🔄 Not Started  
**Priority**: High
**Depends on**: #2

### Learning Objectives
By the end of this module, students will be able to:
1. Implement a complete PyTorch training loop
2. Use Automatic Mixed Precision (AMP) for faster training
3. Implement model checkpointing (save best model)
4. Add validation loop with metrics tracking
5. Setup proper logging for monitoring training progress

### Pomodoro 1: Teaching + Setup (25 min)
- [ ] **Teaching (10 min)**: Training loop components, AMP, checkpointing
- [ ] **Review (5 min)**: src/training/train.py structure
- [ ] **Loss Function Discussion (5 min)**: Why Dice for imbalanced classes?
- [ ] **Metrics Overview (5 min)**: IoU, Dice coefficient preview

### Pomodoro 2: Implementation + Testing (25 min)
- [ ] **Code (15 min)**: Fill in train.py sections
  - TODO 3.1: Setup device (CUDA/MPS/CPU)
  - TODO 3.2: Initialize model, loss, optimizer
  - TODO 3.3: Implement training loop with AMP
  - TODO 3.4: Add validation loop
  - TODO 3.5: Implement checkpoint saving
  - TODO 3.6: Add logging
- [ ] **Test (5 min)**: Run pytest tests/test_module_03_training.py -v
- [ ] **Overfit Test (5 min)**: Verify model can overfit single batch

### Files to Edit
- src/training/train.py

### Test Command
pytest tests/test_module_03_training.py -v

### Teaching Notes
See TEACHING.md#module-3-training-loop for:
- AMP implementation details
- Checkpoint structure and strategy
- Loss function selection guide

### Acceptance Criteria
- [ ] Training loop runs without errors
- [ ] AMP reduces memory usage
- [ ] Checkpoints save model + optimizer + scheduler state
- [ ] Can resume training from checkpoint
- [ ] Loss decreases during overfit test
- [ ] All tests in test_module_03_training.py pass

### Resources
- PyTorch AMP Tutorial: https://pytorch.org/tutorials/recipes/recipes/amp_recipe.html
- Effective Training Tips: https://github.com/vahidk/EffectivePyTorch

---
**Labels**: tutorial, module-3, teaching, coding, testing
